perf_bench_test.go

```diff
@@ -24,127 +24,567 @@ package main
+	for i := 0; i < vmNum; i++ {
+		log.Infof("Shutting down VM %d, images: %s", i, images[i%len(images)])
+		vmIDString := strconv.Itoa(i)
+		message, err := funcPool.RemoveInstance(vmIDString, images[i%len(images)], isSyncOffload)
 		require.NoError(t, err, "Function returned error, "+message)
-	} else {
-		require.Equal(t, resp.IsColdStart, false)
+	}
+}
+
+// Returns a list of image names
+func getImages(t *testing.T) []string {
+	var (
+		images = map[string]string{
+			"helloworld":   "vhiveease/helloworld:var_workload",
+			"chameleon":    "vhiveease/chameleon:var_workload",
+			"pyaes":        "vhiveease/pyaes:var_workload",
+			"image_rotate": "vhiveease/image_rotate:var_workload",
+			"json_serdes":  "vhiveease/json_serdes:var_workload",
+			"lr_serving":   "vhiveease/lr_serving:var_workload",
+			"cnn_serving":  "vhiveease/cnn_serving:var_workload",
+			"rnn_serving":  "vhiveease/rnn_serving:var_workload",
+		}
+		funcs  = strings.Split(*funcNames, ",")
+		result []string
+	)
+
+	for _, funcName := range funcs {
+		imageName, isPresent := images[funcName]
+		require.True(t, isPresent, "Function %s is not supported", funcName)
+		result = append(result, imageName)
+	}
+
+	return result
+}
+
+func dumpMetrics(t *testing.T, metrics []map[string]float64, outfile string) {
+	outFile := getOutFile(outfile)
+
+	f, err := os.OpenFile(outFile, os.O_CREATE|os.O_WRONLY, 0666)
+	require.NoError(t, err, "Failed opening file")
+	defer f.Close()
+
+	headerSet := make(map[string]bool)
+	for _, metric := range metrics {
+		for area := range metric {
+			headerSet[area] = true
+		}
 	}
 
-	log.Debugf("vmID %s: returned in %f seconds", vmIDString, time.Since(start).Seconds())
+	var headers []string
+	for area := range headerSet {
+		headers = append(headers, area)
+	}
+
+	writer := csv.NewWriter(f)
+
+	err = writer.Write(headers)
+	require.NoError(t, err, "Failed writting file")
+	writer.Flush()
+
+	for _, metric := range metrics {
+		var data []string
+		for _, header := range headers {
+			value, isPresent := metric[header]
+			if isPresent {
+				vStr := strconv.FormatFloat(value, 'f', -1, 64)
+				data = append(data, vStr)
+			} else {
+				data = append(data, "")
+			}
+		}
+		err = writer.Write(data)
+		require.NoError(t, err, "Failed writting file")
+		writer.Flush()
+	}
 }
 
-func getImages() map[string]string {
-	return map[string]string{
-		"helloworld": "vhiveease/helloworld:var_workload",
-		// "chameleon":    "vhiveease/chameleon:var_workload",
-		// "pyaes":        "vhiveease/pyaes:var_workload",
-		// "image_rotate": "vhiveease/image_rotate:var_workload",
-		// "json_serdes":  "vhiveease/json_serdes:var_workload",
-		// "lr_serving":   "vhiveease/lr_serving:var_workload",
-		// "cnn_serving":  "vhiveease/cnn_serving:var_workload",
-		// "rnn_serving":  "vhiveease/rnn_serving:var_workload",
-		// "lr_training":  "vhiveease/lr_training:var_workload",
+// getCPUIntenseRPS returns the number of requests per second that stress CPU for each image.
```

Comment:
this comment is not precise. Returns the number of requests completed in a close loop

rename: getCloseLoopRPS

Reply:
Done

Reply:
still not precise. please say exactly how these latencies were measured